File: apps/ios/dakka/Core/Call/ICEServerService.swift
import Foundation
import WebRTC

/// Получение ICE-серверов (STUN + TURN) с бэкенда.
/// Пароль TURN не зашит в бинарь — раздаётся через защищённый endpoint.
enum ICEServerService {

    private struct Server: Decodable {
        let urls: [String]
        let username: String?
        let credential: String?
    }
    private struct Resp: Decodable {
        let iceServers: [Server]
    }

    /// Запрашивает и возвращает массив RTCIceServer.
    /// Fallback: при сетевой ошибке отдаём только Google STUN.
    static func fetch() async -> [RTCIceServer] {
        do {
            let resp: Resp = try await APIClient.shared.get("webrtc/ice-servers")
            return resp.iceServers.map { s in
                if let user = s.username, let cred = s.credential {
                    return RTCIceServer(
                        urlStrings: s.urls,
                        username: user,
                        credential: cred
                    )
                }
                return RTCIceServer(urlStrings: s.urls)
            }
        } catch {
            print("[ICE] fetch failed, using fallback:", error)
            return [
                RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"]),
            ]
        }
    }
}

File: apps/android/app/src/main/java/online/akkdmsg/dakka/data/api/ICEServerService.kt
package online.akkdmsg.dakka.data.api

import kotlinx.serialization.Serializable
import org.webrtc.PeerConnection

object ICEServerService {

    @Serializable
    private data class Server(
        val urls: List<String>,
        val username: String? = null,
        val credential: String? = null,
    )

    @Serializable
    private data class Resp(val iceServers: List<Server>)

    /** Fetch с бэка + fallback на Google STUN при ошибке. */
    suspend fun fetch(): List<PeerConnection.IceServer> = try {
        val resp: Resp = ApiClient.get("webrtc/ice-servers")
        resp.iceServers.map { s ->
            val b = PeerConnection.IceServer.builder(s.urls)
            if (s.username != null) b.setUsername(s.username)
            if (s.credential != null) b.setPassword(s.credential)
            b.createIceServer()
        }
    } catch (_: Exception) {
        listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
        )
    }
}
